### ISSUE TITLE
v4.x: Using --map-by core:span causes error in MPI_Init

### ISSUE THREAD

**Issue by lyu** [open-mpi/ompi#7867]
### What version of Open MPI are you using?
`4.0.4` from release tarball. ~~Can't test the git version since it requires `pandoc`.~~

### Describe how Open MPI was installed
OpenMPI configuration:
`./configure --prefix=$PREFIX --without-verbs --with-hwloc=internal`

### Please describe the system on which you are running
* Operating system/version: CentOS Linux release 7.5.1804
* Computer hardware: Intel Xeon CPU E5-2690 v3
* Network type: Mellanox Technologies MT27600 [Connect-IB]

A few more systems were tested (laptops, ARM machines) and they all have the same issue.

-----------------------------

## Details of the problem

Compile the following MPI program:

```
#include "mpi.h"

int main() {
    MPI_Init(NULL, NULL);
    MPI_Finalize();
}
```

Run it with
`mpirun -n 4 --map-by core:span ./a.out`

Error message:
```
[login2:55588] PMIX ERROR: NOT-FOUND in file dstore_base.c at line 2866
[login2:55588] PMIX ERROR: NOT-FOUND in file server/pmix_server.c at line 3408
[login2:55592] PMIX ERROR: OUT-OF-RESOURCE in file client/pmix_client.c at line 231
[login2:55592] OPAL ERROR: Error in file pmix3x_client.c at line 112
*** An error occurred in MPI_Init
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[login2:55592] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[21660,1],0]
  Exit code:    1
--------------------------------------------------------------------------
[login2:55588] PMIX ERROR: ERROR in file gds_ds21_lock_pthread.c at line 99
[login2:55588] PMIX ERROR: ERROR in file gds_ds21_lock_pthread.c at line 99
```

I tested this on both the login node and the compute node, the error persists as long as I pass `--map-by core:span` or `--map-by numa:span` to `mpirun`.

OpenMPI 4.0.3 works fine.

**Comment by tonycurtis** [comment]
BTW you should be able to just create a shell script that does `exit 0` or equivalent called `pandoc` ( or symlink to `/bin/true` for example).  That will pass.

**Comment by hppritcha** [comment]
I'm able to reproduce this.

**Comment by jsquyres** [comment]
> BTW you should be able to just create a shell script on your PATH that does `exit 0` or equivalent called `pandoc` ( or symlink to `/bin/true` for example). That will pass. (Required by PMIx if memory serves.)

Or you could just download a pandoc binary from https://github.com/jgm/pandoc/releases/ 😄

**Comment by lyu** [comment]
OK, built OMPI from the master branch, the error messages are:
```
$ mpirun -n 4 --map-by core:span ./a.out 
--------------------------------------------------------------------------
A request for multiple cpus-per-proc was given, but a directive
was also given to map to an object level that has less cpus than
requested ones:

  #cpus-per-proc:  1
  number of cpus:  0
  map-by:          UNKNOWN

Please specify a mapping level that has more cpus, or else let us
define a default mapping that will allow multiple cpus-per-proc.
--------------------------------------------------------------------------
```
Not sure what happened, but looks like a different issue, or simply because of changes to launch parameters.

**Comment by jsquyres** [comment]
Yes, master and the v4.x branches have an entirely different back-end for run-time support (i.e., master is what will eventually become v5.0.0, and that has had a complete revamp of the run-time system compared to the v4.0.x/v4.1.x series).

**Comment by hppritcha** [comment]
@jsquyres other binding options also exhibit this behavior

**Comment by jsquyres** [comment]
@hppritcha Mapping or binding or both?  Can you provide a list?

**Comment by hppritcha** [comment]
I'm busy bisecting

**Comment by hppritcha** [comment]
git bisect is implicating 7643 commit 79c426885a4af7e2329e35acfc81ad6f7fc560ff

**Comment by jsquyres** [comment]
What's the path forward on this one?

Is it just `--map-by core:span`, or are other mapping options affected?

Are binding options affected, too?

**Comment by lyu** [comment]
@jsquyres Here is a list of `--map-by` options that can reproduce the error when used on OMPI 4.0.4
```
--map-by hwthread
--map-by core
--map-by numa
--map-by socket
--map-by L1cache
--map-by L2cache
--map-by L3cache
```
Adding modifiers like `:span` or `:PE=n` has no effect, the error persists.
Also, `--rank-by xxx` causes the same error using `xxx` from the list above.
Finally, `--bind-to xxx` has no effect, whether used alone or combined with other options.

**Comment by rhc54** [comment]
The problem is clearly in the rank-by code. When you map-by an object, it automatically sets rank-by to the same object. You can work around the problem for now by simply adding `--rank-by slot` to your cmd line.

**Comment by rhc54** [comment]
FWIW: I cannot replicate any of these problems, including the `--map-by core:span` report for master, once master has been updated by https://github.com/open-mpi/ompi/pull/7886. I haven't looked at v4.

**Comment by rhc54** [comment]
I should note that there is nothing in that PR that I can see as having an impact on the map/rank behavior.

**Comment by jsquyres** [comment]
This is now fixed in the v4.0.x and v4.1.x branches.  Closing.
